libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,271 @@
+
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Dominator analysis of a control flow graph.
+ * The implementation is based on the following paper:
+ * https://www.cs.princeton.edu/courses/archive/spr03/cs423/download/dominators.pdf
+ * See appendix B pg. 139.
+ */
+#pragma once
+
+#include <libyul/backends/evm/ControlFlowGraph.h>
+#include <libsolutil/Visitor.h>
+
+#include <vector>
+#include <map>
+#include <set>
+#include <deque>
+
+namespace solidity::yul
+{
+
+template<typename Vertex, typename ForEachSuccessor>
+class Dominator
+{
+public:
+
+	Dominator(Vertex _entry, size_t _numVertices)
+	{
+		m_vertex = std::vector<Vertex>(_numVertices);
+		m_immediateDominator = lengauerTarjanDominator(_entry, _numVertices);
+		buildDominatorTree();
+	}
+
+	std::vector<Vertex> vertices() const
+	{
+		return m_vertex;
+	}
+
+	std::map<Vertex, size_t> vertexIndices() const
+	{
+		return m_vertexIndex;
+	}
+
+	std::vector<size_t> immediateDominators() const
+	{
+		return m_immediateDominator;
+	}
+
+	std::map<size_t, std::vector<size_t>> dominatorTree() const
+	{
+		return m_dominatorTree;
+	}
```

Comment:
Related to this, I would not assume `Vertex` can be used to index a map or be stored in a `set`. You already use indexes in a lot of your data structures so I'd just go all the way and use them for everything.

Reply:
> Related to this, I would not assume `Vertex` can be used to index a map or be stored in a `set`. You already use indexes in a lot of your data structures so I'd just go all the way and use them for everything.

Yeah, see the following comment regarding why a map from `Vertex` to `size_t` is needed: https://github.com/ethereum/solidity/pull/14460#discussion_r1376166478. We could impose constraints on `Vertex` to make this assumption a requirement, but this would needs C++20.

Reply:
But if you have assumptions about `Vertex`, you should document them. I assumed it could be anything, which is why I was complaining about unnecessary copies here.

The problem here is that it's not that easy to tell what your assumptions about it are just by looking at how you use it. It confused me last time I was reviewing this. On one hand you use `Vertex` for indexing and copy it freely, which means you think of it as a light unique ID to be passed by value. On the other hand, you don't use it like I'd expect an ID to be used at all; you just dump the values a vector and often refer to them indirectly via indices as if they were something heavy. You also don't use them consistently as vertex IDs in class API - for example  `immediateDominators()` and `dominatorTree()` give user the internal numeric IDs instead of referring to vertices using that type.

I think you should commit to one way or the other:
1. If this is a heavy object, store it once, don't make so many copies.
2. If it's just an ID, use it everywhere as a primary way to refer to vertices. Most of the instances of `size_t` should really be `Vertex`. You still need DFS indices in a few places for the algorithm, but get them via `m_vertexIndices`. Also, in this case calling the template parameter `VertexID` would also make things clearer.

Reply:
Done. I'm using IDs everywhere now and DFS index only internally (not exposed to the caller), since most of the operations depends on them. I also added some constraints about the need of a vertex to have a ID. No vertex object is stored in the `DominatorFinder`.